File: src/test/java/no/nav/sosialhjelp/soknad/business/batch/LagringsSchedulerTest.java
package no.nav.sosialhjelp.soknad.business.batch;

import no.nav.sosialhjelp.soknad.business.db.repositories.soknadunderarbeid.BatchSoknadUnderArbeidRepository;
import no.nav.sosialhjelp.soknad.client.leaderelection.LeaderElection;
import no.nav.sosialhjelp.soknad.domain.SoknadUnderArbeid;
import no.nav.sosialhjelp.soknad.domain.SoknadUnderArbeidStatus;
import no.nav.sosialhjelp.soknad.innsending.HenvendelseService;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.mockito.Mock;
import org.mockito.junit.jupiter.MockitoExtension;

import java.util.Arrays;
import java.util.Collections;

import static org.mockito.ArgumentMatchers.anyLong;
import static org.mockito.Mockito.times;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;

@ExtendWith(MockitoExtension.class)
class LagringsSchedulerTest {

    @Mock
    private LeaderElection leaderElection;
    @Mock
    private HenvendelseService henvendelseService;
    @Mock
    private BatchSoknadUnderArbeidRepository batchSoknadUnderArbeidRepository;

    private LagringsScheduler scheduler;

    @BeforeEach
    public void setup() {
        scheduler = new LagringsScheduler(leaderElection, henvendelseService, batchSoknadUnderArbeidRepository, true);

        when(leaderElection.isLeader()).thenReturn(true);
    }

    @Test
    void skalAvbryteIHenvendelseOgSletteFraDatabase() throws InterruptedException {
        String behandlingsId = "2";
        String tilknyttetBehandlingsId = "1";
        long soknadId = 2;
        SoknadUnderArbeid soknadUnderArbeid = new SoknadUnderArbeid()
                .withSoknadId(soknadId)
                .withEier("11111111111")
                .withBehandlingsId(behandlingsId)
                .withStatus(SoknadUnderArbeidStatus.UNDER_ARBEID)
                .withTilknyttetBehandlingsId(tilknyttetBehandlingsId);

        when(batchSoknadUnderArbeidRepository.hentForeldedeEttersendelser()).thenReturn(Arrays.asList(soknadUnderArbeid));

        scheduler.slettForeldedeEttersendelserFraSoknadUnderArbeidDatabase();

        verify(henvendelseService).avbrytSoknad(behandlingsId, true);
        verify(batchSoknadUnderArbeidRepository).slettSoknad(anyLong());
    }

    @Test
    void skalIkkeAvbryteIHenvendelseOgSletteFraDatabaseDersomDetIkkeErEttersendelse() throws InterruptedException {
        String behandlingsId = "2";
        long soknadId = 2;
        SoknadUnderArbeid soknadUnderArbeid = new SoknadUnderArbeid()
                .withSoknadId(soknadId)
                .withEier("11111111111")
                .withBehandlingsId(behandlingsId)
                .withStatus(SoknadUnderArbeidStatus.UNDER_ARBEID) ;

        when(batchSoknadUnderArbeidRepository.hentForeldedeEttersendelser()).thenReturn(Collections.singletonList(soknadUnderArbeid));

        scheduler.slettForeldedeEttersendelserFraSoknadUnderArbeidDatabase();

        verify(henvendelseService, times(0)).avbrytSoknad(behandlingsId, true);
        verify(batchSoknadUnderArbeidRepository, times(0)).slettSoknad(anyLong());
    }

}

File: src/test/kotlin/no/nav/sosialhjelp/soknad/scheduled/LagringsSchedulerTest.kt
package no.nav.sosialhjelp.soknad.scheduled

import io.mockk.every
import io.mockk.just
import io.mockk.mockk
import io.mockk.runs
import io.mockk.verify
import no.nav.sosialhjelp.soknad.db.repositories.soknadunderarbeid.BatchSoknadUnderArbeidRepository
import no.nav.sosialhjelp.soknad.domain.SoknadUnderArbeid
import no.nav.sosialhjelp.soknad.domain.SoknadUnderArbeidStatus
import no.nav.sosialhjelp.soknad.innsending.HenvendelseService
import no.nav.sosialhjelp.soknad.scheduled.leaderelection.LeaderElection
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test

internal class LagringsSchedulerTest {
    private val leaderElection: LeaderElection = mockk()
    private val henvendelseService: HenvendelseService = mockk()
    private val batchSoknadUnderArbeidRepository: BatchSoknadUnderArbeidRepository = mockk()

    private val scheduler = LagringsScheduler(
        leaderElection,
        henvendelseService,
        batchSoknadUnderArbeidRepository,
        batchEnabled = true,
        schedulerDisabled = false
    )

    @BeforeEach
    fun setup() {
        every { leaderElection.isLeader() } returns true
    }

    @Test
    fun skalAvbryteIHenvendelseOgSletteFraDatabase() {
        val behandlingsId = "2"
        val tilknyttetBehandlingsId = "1"
        val soknadId: Long = 2
        val soknadUnderArbeid = SoknadUnderArbeid()
            .withSoknadId(soknadId)
            .withEier("11111111111")
            .withBehandlingsId(behandlingsId)
            .withStatus(SoknadUnderArbeidStatus.UNDER_ARBEID)
            .withTilknyttetBehandlingsId(tilknyttetBehandlingsId)

        every { batchSoknadUnderArbeidRepository.hentForeldedeEttersendelser() } returns listOf(soknadUnderArbeid)
        every { henvendelseService.avbrytSoknad(any(), any()) } just runs
        every { batchSoknadUnderArbeidRepository.slettSoknad(any()) } just runs

        scheduler.slettForeldedeEttersendelserFraSoknadUnderArbeidDatabase()

        verify { henvendelseService.avbrytSoknad(behandlingsId, true) }
        verify { batchSoknadUnderArbeidRepository.slettSoknad(any()) }
    }

    @Test
    fun skalIkkeAvbryteIHenvendelseOgSletteFraDatabaseDersomDetIkkeErEttersendelse() {
        val behandlingsId = "2"
        val soknadId: Long = 2
        val soknadUnderArbeid = SoknadUnderArbeid()
            .withSoknadId(soknadId)
            .withEier("11111111111")
            .withBehandlingsId(behandlingsId)
            .withStatus(SoknadUnderArbeidStatus.UNDER_ARBEID)

        every { batchSoknadUnderArbeidRepository.hentForeldedeEttersendelser() } returns listOf(soknadUnderArbeid)

        scheduler.slettForeldedeEttersendelserFraSoknadUnderArbeidDatabase()

        verify(exactly = 0) { henvendelseService.avbrytSoknad(behandlingsId, true) }
        verify(exactly = 0) { batchSoknadUnderArbeidRepository.slettSoknad(any()) }
    }
}
